clang/lib/AST/ExprConstant.cpp

```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
   return Success(APValue(ResultElements.data(), ResultElements.size()), E);
 }
 
+static bool EvaluateVectorOrLValue(APValue &Result, EvalInfo &Info,
+                                   const Expr *E, const QualType &Type) {
+  if (!Evaluate(Result, Info, E))
+    return false;
+
+  if (Result.isLValue()) {
+    // Source of the data is an lvalue; Manually handle the lvalue as if
+    // it was an rvalue to get the current APValue.
+    LValue LValueFound;
+    LValueFound.setFrom(Info.Ctx, Result);
+    if (!handleLValueToRValueConversion(Info, E, Type, LValueFound, Result))
+      return false;
+  }
+
+  return Result.isVector();
+}
+
+static bool handleVectorConversion(EvalInfo &Info, const FPOptions FPO,
+                                   const Expr *E, QualType SourceTy,
+                                   QualType DestTy, APValue const &Original,
+                                   APValue &Result) {
+  if (SourceTy->isIntegerType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = APValue(APFloat(0.0));
+      return HandleIntToFloatCast(Info, E, FPO, SourceTy, Original.getInt(),
+                                  DestTy, Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(
+          HandleIntToIntCast(Info, E, DestTy, SourceTy, Original.getInt()));
+      return true;
+    }
+  } else if (SourceTy->isRealFloatingType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = Original;
+      return HandleFloatToFloatCast(Info, E, SourceTy, DestTy,
+                                    Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(APSInt());
+      return HandleFloatToIntCast(Info, E, SourceTy, Original.getFloat(),
+                                  DestTy, Result.getInt());
+    }
+  }
+  return false;
```

Comment:
I certainly don't know about any such case...

Reply:
it seems like the only one missing is _BitInt(8) or bigger. I'll just add support for that

Reply:
It looks like the bit-precise support "just worked"? At least, I see a test for it, but not any changes in this function. That's neat!

I'm still leaning towards adding a diagnostic here, though: I'm imagining a future standard or compiler version adding a new vector-capable type (_Complex, maybe? One of the fixed-point types?), and callers of `__builtin_convertvector(...)` with an that type will just see some flavor of "must be initialized by a constant expression," despite  the docs' promise that "This builtin can be used within constant expressions." 

I'm not sure how to test it—maybe using an incomplete type?—but I'm also pretty comfortable with an un-exercised `Info.FFDiag(...)` call here too. That said, this certainly does work with everything I can think to throw at it today, though, so I'm open to calling it a future follow-up idea too. What do you think?

Reply:
I agree that adding a diagnostic here, even if unreachable, is more than just desired, I'm well convinced on the future-proofing argument. I will not add a test because I simply don't have a way to reach this, I believe.